lib/srv/usermgmt_test.go

```diff
@@ -1105,3 +1110,30 @@ func TestHostUsersResolveGroups(t *testing.T) {
 		})
 	}
 }
+
+// errors fetching groups related to a user should not cause panics during UpsertUser
+func TestRegressionGroupErrorDoesNotPanic(t *testing.T) {
+	t.Parallel()
+
+	allGroups := []string{"foo", "bar", "baz"}
+	users, backend := initBackend(t, allGroups)
+
+	userinfo := services.HostUsersInfo{
+		Groups: slices.Clone(allGroups[:2]),
+		Mode:   services.HostUserModeKeep,
+	}
+
+	// Create user
+	closer, err := users.UpsertUser("alice", userinfo)
+	assert.NoError(t, err)
+	assert.Equal(t, nil, closer)
+	assert.Zero(t, backend.setUserGroupsCalls)
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportKeepGroup), backend.users["alice"])
+	assert.NotContains(t, backend.users["alice"], types.TeleportDropGroup)
+
+	backend.groupDatabaseErr = errors.New("could not find group")
+	require.NotPanics(t, func() {
+		closer, err = users.UpsertUser("alice", userinfo)
+	})
```

Comment:
I'm not sure this is how we should be testing this. You can make the argument that none of our code should panic (unless explicitly coded to do so), in which case we would have every test include a `requires.NotPanics()` wrapped around everything.

In this case, there was a panic due to mistaken error handling. Which says to me we need to test that the error handling is now correct, which means `UpsertUser()` returns an error when the group lookup returns an error. It is sufficient to just have the next line (`require.Error(t, err)`) and to remove the `NotPanic` assertion, as if it does panic, the test will still fail as panics in the code under test will always fail a test unless specifically caught and ignored.

Reply:
Spot on, Cam.

Reply:
Great point. I included the `require.NotPanic()` just capture the specific regression this was guarding against, but you're right that it isn't actually testing anything valuable :+1: Updated!